src/Omega_h_meshsim.cpp

```diff
@@ -220,102 +385,213 @@ void read_internal(pMesh m, Mesh* mesh) {
-  auto vert2quad = mesh->ask_up(Topo_type::vertex, Topo_type::quadrilateral);
+  Adj tri2vert;
+  Adj vert2tri;
+  Adj quad2vert;
+  Adj vert2quad;
+  if (is_simplex || is_hypercube) {
+    tri2vert = mesh->ask_down(2, 0);
+    vert2tri = mesh->ask_up(0, 2);
+  }
+  else {
+    tri2vert = mesh->ask_down(Topo_type::triangle, Topo_type::vertex);
+    vert2tri = mesh->ask_up(Topo_type::vertex, Topo_type::triangle);
+    quad2vert = mesh->ask_down(Topo_type::quadrilateral, Topo_type::vertex);
+    vert2quad = mesh->ask_up(Topo_type::vertex, Topo_type::quadrilateral);
+  }
+  printf("ok4\n");
 
   HostWrite<LO> host_tet2verts(count_tet*4);
   for (Int i = 0; i < count_tet; ++i) {
     for (Int j = 0; j < 4; ++j) {
       host_tet2verts[i*4 + j] =
-          elem_vertices[0][static_cast<std::size_t>(i*4 + j)];
+          rgn_vertices[0][static_cast<std::size_t>(i*4 + j)];
     }
   }
   auto tet2verts = Read<LO>(host_tet2verts.write());
-  down = reflect_down(tet2verts, tri2vert.ab2b, vert2tri, Topo_type::tetrahedron, Topo_type::triangle);
-  mesh->set_ents(Topo_type::tetrahedron, Topo_type::triangle, down);
+  if (is_simplex) {
+    down = reflect_down(tet2verts, tri2vert.ab2b, vert2tri,
+                        OMEGA_H_SIMPLEX, 3, 2);
+    mesh->set_ents(3, down);
+    mesh->add_tag<ClassId>(3, "class_id", 1,
+                           Read<ClassId>(host_class_ids_rgn.write()));
+    mesh->add_tag<I8>(3, "class_dim", 1,
+                      Read<I8>(host_class_dim_rgn.write()));
+  }
+  else if (is_hypercube) {
+  }
+  else {
+    down = reflect_down(tet2verts, tri2vert.ab2b, vert2tri,
+                        Topo_type::tetrahedron, Topo_type::triangle);
+    mesh->set_ents(Topo_type::tetrahedron, Topo_type::triangle, down);
+    mesh->add_tag<ClassId>(Topo_type::tetrahedron, "class_id", 1,
+                           Read<ClassId>(host_class_ids_tet.write()));
+    mesh->add_tag<I8>(Topo_type::tetrahedron, "class_dim", 1,
+                      Read<I8>(host_class_dim_tet.write()));
+  }
+  printf("ok5\n");
 
   HostWrite<LO> host_hex2verts(count_hex*8);
   for (Int i = 0; i < count_hex; ++i) {
     for (Int j = 0; j < 8; ++j) {
       host_hex2verts[i*8 + j] =
-          elem_vertices[1][static_cast<std::size_t>(i*8 + j)];
+          rgn_vertices[1][static_cast<std::size_t>(i*8 + j)];
     }
   }
   auto hex2verts = Read<LO>(host_hex2verts.write());
-  down = reflect_down(hex2verts, quad2vert.ab2b, vert2quad, Topo_type::hexahedron, Topo_type::quadrilateral);
-  mesh->set_ents(Topo_type::hexahedron, Topo_type::quadrilateral, down);
+  if (is_hypercube) {
+    down = reflect_down(hex2verts, quad2vert.ab2b, vert2quad,
+                        OMEGA_H_HYPERCUBE, 3, 2);
+    mesh->set_ents(3, down);
+    mesh->add_tag<ClassId>(3, "class_id", 1,
+                           Read<ClassId>(host_class_ids_rgn.write()));
+    mesh->add_tag<I8>(3, "class_dim", 1,
+                      Read<I8>(host_class_dim_rgn.write()));
+  }
+  else if (is_simplex) {
```

Comment:
We should add a comment for future readers; the empty conditional is a bit puzzling at first sight.